src/graph/UpdateEdgeExecutor.cpp

```diff
@@ -0,0 +1,258 @@
+                    default:
+                        LOG(FATAL) << "Unknown VariantType: " << column.which();
+                }
+            }
+        }
+        rows.emplace_back();
+        rows.back().set_columns(std::move(row));
+    }
+    resp_->set_rows(std::move(rows));
+    DCHECK(onFinish_);
+    onFinish_();
+}
+
+
+void UpdateEdgeExecutor::setupResponse(cpp2::ExecutionResponse &resp) {
+    if (resp_ == nullptr) {
+        resp_ = std::make_unique<cpp2::ExecutionResponse>();
+    }
+    resp = std::move(*resp_);
+}
+
+
+void UpdateEdgeExecutor::insertReverselyEdge() {
+    std::vector<storage::cpp2::Edge> edges;
+    storage::cpp2::Edge reverselyEdge;
+    reverselyEdge.key.set_src(edge_.dst);
+    reverselyEdge.key.set_dst(edge_.src);
+    reverselyEdge.key.set_ranking(edge_.ranking);
+    reverselyEdge.key.set_edge_type(-edge_.edge_type);
+    reverselyEdge.props = "";
+    edges.emplace_back(reverselyEdge);
+    auto space = ectx()->rctx()->session()->space();
+    auto future = ectx()->storage()->addEdges(space, std::move(edges), false);
+    auto *runner = ectx()->rctx()->runner();
+    auto cb = [this] (auto &&resp) {
+        auto completeness = resp.completeness();
+        if (completeness != 100) {
+            // Very bad, it should delete the upsert positive edge!!!
+            DCHECK(onError_);
+            onError_(Status::Error("Insert the reversely edge failed."));
+            return;
+        }
+    };
+    auto error = [this] (auto &&e) {
+        LOG(ERROR) << "Exception caught: " << e.what();
+        // Very bad, it should delete the upsert positive edge!!!
+        DCHECK(onError_);
+        onError_(Status::Error("Internal error: insert reversely edge."));
+    };
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+
+void UpdateEdgeExecutor::execute() {
+    FLOG_INFO("Executing UpdateEdge: %s", sentence_->toString().c_str());
+    auto space = ectx()->rctx()->session()->space();
+    std::string filterStr = filter_ ? Expression::encode(filter_) : "";
+    auto returns = getReturnColumns();
+    auto future = ectx()->storage()->updateEdge(space,
+                                                edge_,
+                                                filterStr,
+                                                std::move(updateItems_),
+                                                std::move(returns),
+                                                insertable_);
+    auto *runner = ectx()->rctx()->runner();
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            DCHECK(onError_);
+            onError_(std::move(resp).status());
+            return;
+        }
+        auto rpcResp = std::move(resp).value();
+        if (insertable_ && rpcResp.get_upsert()) {
+            // TODO(zhangguoqing) Making the reverse edge of insertion is transactional
+            this->insertReverselyEdge();
+        }
+        this->finishExecution(std::move(rpcResp));
```

Comment:
can not call finishExecution after insertReverselyEdge.

Reply:
done!